File: ballast-core/npm/src/domain/events/player-signed-out.ts
import { EventBase } from "../../messaging/event-base";
import { IDomainEvent } from "../domain-event";
import { Player } from "../models/player";

export class PlayerSignedOutDomainEvent extends EventBase implements IDomainEvent {

    public static readonly id: string = "PlayerSignedOutDomainEvent";
    public get id() {
        return PlayerSignedOutDomainEvent.id;
    }

    public readonly player: Player | null; 

    private constructor(eventDateIsoString: string, player: Player | null = null) {
        super(eventDateIsoString);
        this.player = player || null;
    }

    public static fromPlayer(player: Player | null = null) {
        return new PlayerSignedOutDomainEvent(
            EventBase.getDateIsoString(),
            player
        );
    }

}

File: ballast-core/dotnet/src/application/events/PlayerSignedOut.cs
using Ballast.Core.Application.Models;
using Ballast.Core.Messaging;

namespace Ballast.Core.Application.Events
{
    public class PlayerSignedOutEvent : EventBase 
    {

        public static string GetId() => nameof(PlayerSignedOutEvent);
        public override string Id => GetId();

        public PlayerDto Player { get; private set; }

        private PlayerSignedOutEvent() { } // For model binding / deserialization
        private PlayerSignedOutEvent(string eventDateIsoString, PlayerDto player = null) : base(eventDateIsoString) 
        {
            Player = player;
        }

        public static PlayerSignedOutEvent FromPlayer(PlayerDto player = null) =>
            new PlayerSignedOutEvent(
                EventBase.GetDateIsoString(),
                player
            );

    }
}